T7655: Change VPP memory verification

natali-rs1985 opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Other (please describe): 

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
<!-- * https://vyos.dev/Txxxx -->
 * https://vyos.dev/T7655
## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->

## Proposed changes
<!--- Describe your changes in detail -->

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->
```
vyos@vyos# python3 /usr/libexec/vyos/tests/smoke/cli/test_vpp.py
test_01_vpp_basic (__main__.TestVPP.test_01_vpp_basic) ... ok
test_02_vpp_vxlan (__main__.TestVPP.test_02_vpp_vxlan) ... ok
test_03_vpp_gre (__main__.TestVPP.test_03_vpp_gre) ... ok
test_04_vpp_geneve (__main__.TestVPP.test_04_vpp_geneve) ... skipped 'Skipping this test geneve index always is 0'
test_05_vpp_loopback (__main__.TestVPP.test_05_vpp_loopback) ... ok
test_06_vpp_bonding (__main__.TestVPP.test_06_vpp_bonding) ... skipped 'Skipping temporary bonding, sometimes get recursion T7117'
test_07_vpp_bridge (__main__.TestVPP.test_07_vpp_bridge) ... ok
test_08_vpp_ipip (__main__.TestVPP.test_08_vpp_ipip) ... ok
test_09_vpp_xconnect (__main__.TestVPP.test_09_vpp_xconnect) ... ok
test_10_vpp_driver_options (__main__.TestVPP.test_10_vpp_driver_options) ... ok
test_11_vpp_cpu_settings (__main__.TestVPP.test_11_vpp_cpu_settings) ... ok
test_12_vpp_cpu_corelist_workers (__main__.TestVPP.test_12_vpp_cpu_corelist_workers) ... ok
test_13_1_buffer_page_size (__main__.TestVPP.test_13_1_buffer_page_size) ... ok
test_13_2_statseg_page_size (__main__.TestVPP.test_13_2_statseg_page_size) ... ok
test_13_3_mem_page_size (__main__.TestVPP.test_13_3_mem_page_size) ... skipped 'Skipping'
test_14_mem_default_hugepage (__main__.TestVPP.test_14_mem_default_hugepage) ... ok
test_15_vpp_ipsec_xfrm_nl (__main__.TestVPP.test_15_vpp_ipsec_xfrm_nl) ... ok
test_16_vpp_cgnat (__main__.TestVPP.test_16_vpp_cgnat) ... ok
test_17_vpp_nat (__main__.TestVPP.test_17_vpp_nat) ... ok

----------------------------------------------------------------------
Ran 19 tests in 566.917s

OK (skipped=3)
```
## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

sever-sever commented on python/vyos/vpp/config_resource_checks/memory.py:
```diff
@@ -45,7 +45,7 @@ def get_total_hugepages_free_memory() -> int:
     """
     info = get_hugepages_info()
-    hugepages_free = info.get('HugePages_Free')
+    hugepages_free = info.get('HugePages_Total')
```
The variable name is `hugepages_free`, but you get info `HugePages_Total`
The variable name or key should be changed.

natali-rs1985 replied:
changed, thanks

natali-rs1985 committed:
T7655: Change VPP memory verification

Get Hugepages information from /sys/kernel/mm/hugepages instead of
/proc/meminfo
Add an error margin for memory calculations

sever-sever commented on python/vyos/vpp/config_verify.py:
```diff
@@ -226,27 +226,28 @@ def verify_vpp_memory(config: dict):
-        memory_required -= memory_used
-
-    if memory_required > available_memory:
+    if main_heap_size < available_memory:
```
VPP cannot start and smoketst fails
```
vyos@r14# run show conf com | match kernel
set system option kernel memory default-hugepage-size '2M'
set system option kernel memory hugepage-size 2M hugepage-count '2048'
[edit]
vyos@r14# 
```
Add VPP config:
```
vyos@r14# show vpp
+settings {
+    interface eth1 {
+        driver dpdk
+    }
+    unix {
+        poll-sleep-usec 222
+    }
+}
[edit]
vyos@r14# commit
[ vpp ]
"memory main-heap-size" must not be greater than hugepages memory.
Reduce to 3072M or less
[[vpp]] failed
Commit failed
[edit]
vyos@r14# 

```

sever-sever replied:
In debug mode with `pprint`
```
vyos@r14# commit
[ vpp ]
{'kernel_memory_settings': {'default_hugepage_size': '2M',
                            'hugepage_size': {'2M': {'hugepage_count': '2048'}}},
 'nat': {'cgnat': {'timeout': {'icmp': '60',
                               'tcp_established': '7440',
                               'tcp_transitory': '240',
                               'udp': '300'}}},
 'persist_config': {'eth1': {'bus_id': 'virtio',
                             'dev_id': '0000:07:00.0',
                             'original_driver': 'virtio_net'}},
 'settings': {'host_resources': {'max_map_count': '65535',
                                 'shmmax': '2147483648'},
              'interface': {'eth1': {'dpdk_options': {'dev_id': '0000:07:00.0'},
                                     'driver': 'dpdk',
                                     'kernel_module': 'virtio_net'}},
              'ipv6': {'hash_buckets': '65536', 'heap_size': '32M'},
              'l2learn': {'limit': '4194304'},
              'memory': {'default_hugepage_size': '2M', 'main_heap_size': '3G'},
              'nat44': {'session_limit': '64512',
                        'timeout': {'icmp': '60',
                                    'tcp_established': '7440',
                                    'tcp_transitory': '240',
                                    'udp': '300'}},
              'unix': {'poll_sleep_usec': '222'}}}
"memory main-heap-size" must not be greater than hugepages memory.
Reduce to 3072M or less
[[vpp]] failed
Commit failed
[edit]
vyos@r14# 
```

natali-rs1985 replied:
Fixed.

sever-sever reviewed (approved):
All smoketests and manual tests work fine:
```
vyos@r14:~$ /usr/libexec/vyos/tests/smoke/cli/test_vpp.py
test_01_vpp_basic (__main__.TestVPP.test_01_vpp_basic) ... ok
test_02_vpp_vxlan (__main__.TestVPP.test_02_vpp_vxlan) ... ok
test_03_vpp_gre (__main__.TestVPP.test_03_vpp_gre) ... ok
test_04_vpp_geneve (__main__.TestVPP.test_04_vpp_geneve) ... skipped 'Skipping this test geneve index always is 0'
test_05_vpp_loopback (__main__.TestVPP.test_05_vpp_loopback) ... ok
test_06_vpp_bonding (__main__.TestVPP.test_06_vpp_bonding) ... skipped 'Skipping temporary bonding, sometimes get recursion T7117'
test_07_vpp_bridge (__main__.TestVPP.test_07_vpp_bridge) ... ok
test_08_vpp_ipip (__main__.TestVPP.test_08_vpp_ipip) ... ok
test_09_vpp_xconnect (__main__.TestVPP.test_09_vpp_xconnect) ... ok
test_10_vpp_driver_options (__main__.TestVPP.test_10_vpp_driver_options) ... ok
test_11_vpp_cpu_settings (__main__.TestVPP.test_11_vpp_cpu_settings) ... ok
test_12_vpp_cpu_corelist_workers (__main__.TestVPP.test_12_vpp_cpu_corelist_workers) ... ok
test_13_1_buffer_page_size (__main__.TestVPP.test_13_1_buffer_page_size) ... ok
test_13_2_statseg_page_size (__main__.TestVPP.test_13_2_statseg_page_size) ... ok
test_13_3_mem_page_size (__main__.TestVPP.test_13_3_mem_page_size) ... ok
test_14_mem_default_hugepage (__main__.TestVPP.test_14_mem_default_hugepage) ... ok
test_15_vpp_ipsec_xfrm_nl (__main__.TestVPP.test_15_vpp_ipsec_xfrm_nl) ... ok
test_16_vpp_cgnat (__main__.TestVPP.test_16_vpp_cgnat) ... ok
test_17_vpp_nat (__main__.TestVPP.test_17_vpp_nat) ... ok
test_18_vpp_sflow (__main__.TestVPP.test_18_vpp_sflow) ... ok

----------------------------------------------------------------------
Ran 20 tests in 355.336s

OK (skipped=2)
vyos@r14:~$ 
```
